src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -1181,91 +1181,100 @@ void ShenandoahHeap::cancel_old_gc() {
   }
 }
 
-// xfer_limit is the maximum we're able to transfer from young to old
+// Make sure old-generation is large enough, but no larger than is necessary, to hold mixed evacuations
+// and promotions, if we anticipate either. Any deficit is provided by the young generation, subject to
+// xfer_limit, and any excess is transferred to the young generation.
+// xfer_limit is the maximum we're able to transfer from young to old.
 void ShenandoahHeap::adjust_generation_sizes_for_next_cycle(
   size_t xfer_limit, size_t young_cset_regions, size_t old_cset_regions) {
 
-  // Make sure old-generation is large enough, but no larger, than is necessary to hold mixed evacuations
-  // and promotions if we anticipate either.
-  size_t region_size_bytes = ShenandoahHeapRegion::region_size_bytes();
-  size_t promo_load = get_promotion_potential();
+  // We can limit the old reserve to the size of anticipated promotions:
+  // max_old_reserve is an upper bound on memory evacuated from old and promoted to old,
+  // clamped by the old generation space available.
+  //
+  // Here's the algebra.
+  // Let SOEP = ShenandoahOldEvacRatioPercent,
+  //     OE = old evac,
+  //     YE = young evac, and
+  //     TE = total evac = OE + YE
+  // By definition:
+  //            SOEP/100 = OE/TE
+  //                     = OE/(OE+YE)
+  //  => SOEP/(100-SOEP) = OE/((OE+YE)-OE)
+  //                     = OE/YE
+  //  =>              OE = YE*SOEP/(100-SOEP)
+
```

Comment:
Same comment as above, regarding the algebra steps.

Reply:
Done.